src/org/nfrac/comportex/cells.cljc

```diff
@@ -841,196 +905,209 @@
 (defrecord LayerActiveState
     [in-ff-bits in-stable-ff-bits
      out-ff-bits out-stable-ff-bits
-     col-overlaps matching-ff-seg-paths well-matching-ff-seg-paths
+     col-overlaps matching-ff-seg-paths
      temporal-pooling-exc
-     active-cols burst-cols active-cells col-winners timestep])
+     active-cols burst-cols col-active-cells active-cells timestep])
+
+(defrecord LayerLearnState
+    [col-winners winner-seg learning-cells learning punishments
+    prior-active-cells timestep])
 
 (defrecord LayerDistalState
-    [on-bits on-lc-bits cell-exc pred-cells
-     matching-seg-paths well-matching-seg-paths
-     prior-active-cells timestep])
+    [active-bits learnable-bits cell-exc pred-cells matching-seg-paths timestep])
 
 (def empty-active-state
   (map->LayerActiveState
-   {:col-winners {}
-    :active-cells #{}
+   {:active-cells #{}
     :active-cols #{}
     :temporal-pooling-exc {}
-    :well-matching-ff-seg-paths {}}))
+    :matching-ff-seg-paths {}}))
+
+(def empty-learn-state
+  (map->LayerLearnState
+   {:col-winners {}
+    :learning-cells #{}
+    :learning {}
+    :punishments {}
+    :prior-active-cells #{}}))
 
 (def empty-distal-state
   (map->LayerDistalState
-   {:on-bits #{}
+   {:active-bits #{}
     :cell-exc {}
     :pred-cells #{}
-    :well-matching-seg-paths {}}))
+    :matching-seg-paths {}}))
+
+(defn compute-active-state
+  [state ff-bits stable-ff-bits proximal-sg distal-state apical-state
+   boosts topology inh-radius spec engaged?]
+  (let [;; proximal excitation as number of active synapses, keyed by [col 0 seg-idx]
+        col-seg-overlaps (p/excitations proximal-sg ff-bits
+                                        (:stimulus-threshold (:proximal spec)))
+        ;; these both keyed by [col 0]
+        [raw-col-exc ff-seg-paths]
+        (best-segment-excitations-and-paths col-seg-overlaps)
+        col-exc (cond-> raw-col-exc
+                  (not engaged?)
+                  (select-keys (keys ff-seg-paths))
```

Comment:
Does this still make sense? I'd expect this to be a no-op, since both of these values come from the line above. Previously this filtered out the non-"good" excitations.